### PR TITLE
[FLINK-17273][runtime] ActiveResourceManager closes task manager connection on worker terminated.

### PR DESCRIPTION
**Pull request by xintongsong**
## What is the purpose of the change

This PR fix the problem that active RMs does not close the task manager connection on worker terminated. Without this fix, RM has to wait for heartbeat timeout to find out that TM has terminated.

## Verifying this change

* Add `ActiveResourceManagerTest#testCloseTaskManagerConnectionOnWorkerTerminated`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

**Comment by xintongsong** [comment]
cc @wangyang0918

File: flink-runtime/src/test/java/org/apache/flink/runtime/resourcemanager/active/ActiveResourceManagerTest.java [hunk 1]
```diff
@@ -307,6 +308,33 @@ public void testWorkerTerminatedNoLongerRequired() throws Exception {
...
+
+			driverBuilder.setRequestResourceFunction(taskExecutorProcessSpec -> {
+				requestWorkerFromDriverFuture.complete(taskExecutorProcessSpec);
+				return CompletableFuture.completedFuture(tmResourceId);
+			});
+
+			runTest(() -> {
+				// request a new worker, terminate it after registered
+				runInMainThread(() -> getResourceManager().startNewWorker(WORKER_RESOURCE_SPEC))
+					.thenCompose((ignrore) -> registerTaskExecutor(tmResourceId, taskExecutorGateway))
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/resourcemanager/active/ActiveResourceManagerTest.java:330]:
```suggestion
					.thenCompose((ignore) -> registerTaskExecutor(tmResourceId, taskExecutorGateway))
```

File: flink-runtime/src/test/java/org/apache/flink/runtime/resourcemanager/active/ActiveResourceManagerTest.java [hunk 2]
```diff
@@ -307,6 +308,33 @@ public void testWorkerTerminatedNoLongerRequired() throws Exception {
...
+					.thenRun(() -> runInMainThread(() -> getResourceManager().onWorkerTerminated(tmResourceId, new Exception())));
+				// verify task manager connection is closed
+				disconnectResourceManagerFuture.get(TIMEOUT_SEC, TimeUnit.SECONDS);
+			});
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/resourcemanager/active/ActiveResourceManagerTest.java:334]: I really like this test setup. It is very elegant :-)

File: flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/active/ActiveResourceManager.java [hunk 3]
```diff
@@ -197,11 +197,12 @@ public void onPreviousAttemptWorkersRecovered(Collection<WorkerType> recoveredWo
 	}
 
 	@Override
-	public void onWorkerTerminated(ResourceID resourceId) {
-		log.info("Worker {} is terminated.", resourceId.getStringWithMetadata());
+	public void onWorkerTerminated(ResourceID resourceId, Exception cause) {
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/active/ActiveResourceManager.java:200]: If we don't really use `Exception cause` here, then I would suggest to simply change it into a `String diagnostics` parameter. We can later still change it if we need an `Exception` to be reported.

**Review by tillrohrmann — APPROVED** [review]
Thanks for creating this PR @xintongsong. LGTM. +1 for merging.

**Comment by xintongsong** [comment]
Thanks for the review, @tillrohrmann

**Pull request closed without merging**